In [ ]:
# mount drive
#from google.colab import drive
#drive.mount('/content/drive')

In [ ]:
# move directory
#import os
#colab_dir = "./drive/MyDrive/"
#os.chdir(colab_dir)

In [1]:
# import libraries
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import matplotlib_inline
%matplotlib inline
import seaborn as sns

In [2]:
##  pandas profiling
#!pip install https://github.com/pandas-profiling/pandas-profiling/archive/master.zip    
from pandas_profiling import ProfileReport

In [3]:
# set random seed
import random
random.seed(335)

In [4]:
# magic word
pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 500)

In [5]:
# for better viz
import pprint
import warnings
warnings.filterwarnings('ignore')

In [7]:
# read csv file
#df = pd.read_csv(path + file_name)

### reference
-------------------

- [pandas cheat sheet](https://github.com/pandas-dev/pandas/tree/master/doc/cheatsheet)
- [numpy cheat sheet(data camp)](https://www.datacamp.com/community/blog/python-numpy-cheat-sheet)
- [scikit-learn cheat sheet(data camp)](datacamp.com/community/blog/scikit-learn-cheat-sheet)

# data understanding
------------
The data understanding phase starts with an initial data collection and proceeds withactivities in order to get familiar with the data, to identify data quality problems, to discover first insights into the data or to detect interesting subsets to form hypothesesfor hidden information.

## collect initial data
----------

### task

Acquire within the project the data (or access to the data) listed in the
project resources. This initial collection includes data loading if necessary
for data understanding. For example, if you apply a specific tool for data
understanding, it makes perfect sense to load your data into this tool.
This effort possibly leads to initial data preparation steps.

Note: if you acquire multiple data sources, integration is an additional
issue, either here or in the later data preparation phase.

### output

List the dataset (or datasets) acquired, together with their locations
within the project, the methods used to acquire them and any problems
encountered. Record problems encountered and any solutions achieved
to aid with future replication of this project or with the execution of
similar future projects.

## describe data
----------

### task

Examine the “gross” or “surface” properties of the acquired data and
report on the results.

### output

Describe the data which has been acquired, including: the format of
the data, the quantity of data, for example number of records and fields
in each table, the identities of the fields and any other surface features
of the data which have been discovered. Does the data acquired satisfy
the relevant requirements?

## explore data
----------

### task

This task tackles the data mining questions, which can be addressed
using querying, visualization and reporting. These include: distribution
of key attributes, for example the target attribute of a prediction task;
relations between pairs or small numbers of attributes; results of
simple aggregations; properties of significant sub-populations; simple
statistical analyses. These analyses may address directly the data mining goals; they may also contribute to or refine the data description
and quality reports and feed into the transformation and other data
preparation needed for further analysis.

### output

Describe results of this task including first findings or initial hypothesis and their impact on the remainder of the project. If appropriate,
include graphs and plots, which indicate data characteristics or lead
to interesting data subsets for further examination.

## verify data quality
----------

### task

Examine the quality of the data, addressing questions such as: is the
data complete (does it cover all the cases required)? Is it correct or
does it contain errors and if there are errors how common are they?
Are there missing values in the data? If so how are they represented,
where do they occur and how common are they?

### output

List the results of the data quality verification; if quality problems
exist, list possible solutions. Solutions to data quality problems
generally depend heavily on both data and business knowledge.

# note/questions
-------------

#### collect initial data


#### describe data



#### explore data



#### verify data quality




### Collect initial data
-------------

For this project I've used Online Retail dataset.

It's located on directory: `./data/Online Retail.xlsx`
This file is exel sheet, so it was taken with `read_excel` method with following parameters:

In [6]:
# data location:
path = './data/'
file_name = 'Online Retail.xlsx'
sheet_name = 'Online Retail'

In [7]:
# read excel file
df = pd.read_excel(path + file_name, sheet_name=sheet_name)

### Describe data
----------
As code slippeds shows dataset has following properties:
    
* dataset has one table

* dataset has shape `(541909, 8)` (541909 records and 8 columns)
    
*dataset has follwing columns: `'InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'UnitPrice', 'CustomerID', 'Country`
    
* dataset has follwing formats: `object`, `datetime` (which should be converted to numerical datatype) and `float64`

Now we may conclude that dataset contain information about prices, product name, invoice date, quantity and information about country. Thus given dataset can be used for our problem
    

In [8]:
# number of records and fields
df.shape # (records, fields)

(541909, 8)

In [9]:
df.columns

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country'],
      dtype='object')

In [12]:
# head
df.head(n=5)

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

In [13]:
# tail
df.tail(n=5)

InvoiceNo StockCode                      Description  Quantity  \
541904    581587     22613      PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899     CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254    CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255  CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138    BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID Country  
541904 2011-12-09 12:50:00       0.85     12680.0  France  
541905 2011-12-09 12:50:00       2.10     12680.0  France  
541906 2011-12-09 12:50:00       4.15     12680.0  France  
541907 2011-12-09 12:50:00       4.15     12680.0  France  
541908 2011-12-09 12:50:00       4.95     12680.0  France

During analizing stockcodes it was found that same stoke code may have different `Description` or title of product (example bellow)

Also `Description` may be used to change stoke code example: **wrongly marked carton 22804**

In [23]:
df[df['StockCode']==df['StockCode'].unique()[0]]['Description'].unique()

array(['WHITE HANGING HEART T-LIGHT HOLDER', '?',
       'wrongly marked carton 22804',
       'CREAM HANGING HEART T-LIGHT HOLDER'], dtype=object)

From code bellow we may conclude that StockCodes: 

`['D', 'M', 'S', 'm', 'B', 'C2', 'DOT', 'POST', 'PADS', 'CRUK', 'AMAZONFEE','BANK CHARGES', 'gift_0001_40', 'gift_0001_50', 'gift_0001_30', 'gift_0001_20', 'gift_0001_10']`

should be removed. As this code doesn't describe sold product

In [82]:
letters = []
length = []
examples = {1:[],2:[],3:[],4:[],5:[],6:[],7:[],8:[],9:[],10:[],11:[],12:[]}
def get_letter(code,letters):
    
    for l in code:
        if str(l).isalpha() and not (l in letters):
            letters.append(l)
    
            
for code in df['StockCode']:
    if not len(str(code)) in length:
        length.append(len(str(code)))
    
    if not code in examples[len(str(code))]:
        examples[len(str(code))].append(code)
        
        
    if not str(code).isdigit():
        get_letter(code,letters)
print(letters)
print(length)

for k in examples.keys():
    print(f"length {k}")
    if len(examples[k])<100:
        print('full',examples[k])
    else:
        print(examples[k][:6])
        print("with letters")

['A', 'B', 'G', 'E', 'P', 'O', 'S', 'T', 'L', 'C', 'N', 'D', 'F', 'H', 'M', 'J', 'R', 'K', 'V', 'W', 'b', 'l', 'p', 'a', 's', 'c', 'e', 'I', 'Y', 'd', 'f', 'U', 'Z', 'n', 'k', 'g', 'i', 't', 'm']
[6, 5, 4, 7, 1, 2, 3, 12, 9, 8]
length 1
full ['D', 'M', 'S', 'm', 'B']
length 2
full ['C2']
length 3
full ['DOT']
length 4
full ['POST', 'PADS', 'CRUK']
length 5
[71053, 22752, 21730, 22633, 22632, 84879]
with letters
length 6
['85123A', '84406B', '84029G', '84029E', '82494L', '85099C']
with letters
length 7
full ['15056BL', '15056bl', '79323GR', '79323LP']
length 8
full ['DCGS0076', 'DCGS0003', 'DCGS0070', 'DCGS0055', 'DCGS0072', 'DCGS0074', 'DCGS0069', 'DCGS0057', 'DCGSSBOY', 'DCGS0004', 'DCGS0073', 'DCGS0071', 'DCGS0068', 'DCGS0067']
length 9
full ['AMAZONFEE', 'DCGSSGIRL', 'DCGS0066P']
length 10
full []
length 11
full []
length 12
full ['BANK CHARGES', 'gift_0001_40', 'gift_0001_50', 'gift_0001_30', 'gift_0001_20', 'gift_0001_10']


In [60]:
df[df['StockCode'] == 'gift_0001_40']

InvoiceNo     StockCode                         Description  Quantity  \
38248     539492  gift_0001_40  Dotcomgiftshop Gift Voucher £40.00         1   
145463    548893  gift_0001_40  Dotcomgiftshop Gift Voucher £40.00         1   
281008    561513  gift_0001_40  Dotcomgiftshop Gift Voucher £40.00         1   

               InvoiceDate  UnitPrice  CustomerID         Country  
38248  2010-12-20 10:14:00      34.04         NaN  United Kingdom  
145463 2011-04-04 15:54:00      33.33         NaN  United Kingdom  
281008 2011-07-27 15:12:00      33.33         NaN  United Kingdom

In [81]:
for i in ['D', 'M', 'S', 'm', 'B']:
    print(df[df['StockCode'] == i].head(1))

    InvoiceNo StockCode Description  Quantity         InvoiceDate  UnitPrice  \
141   C536379         D    Discount        -1 2010-12-01 09:41:00       27.5   

     CustomerID         Country  
141     14527.0  United Kingdom  
     InvoiceNo StockCode Description  Quantity         InvoiceDate  UnitPrice  \
2239    536569         M      Manual         1 2010-12-01 15:35:00       1.25   

      CustomerID         Country  
2239     16274.0  United Kingdom  
      InvoiceNo StockCode Description  Quantity         InvoiceDate  \
14436   C537581         S     SAMPLES        -1 2010-12-07 12:03:00   

       UnitPrice  CustomerID         Country  
14436      12.95         NaN  United Kingdom  
      InvoiceNo StockCode Description  Quantity         InvoiceDate  \
40383    539736         m      Manual         1 2010-12-21 15:18:00   

       UnitPrice  CustomerID         Country  
40383       2.55         NaN  United Kingdom  
       InvoiceNo StockCode      Description  Quantity         In

In [59]:
df[df['StockCode'] == 'C2'].head(1)

InvoiceNo StockCode Description  Quantity         InvoiceDate  UnitPrice  \
1423    536540        C2    CARRIAGE         1 2010-12-01 14:05:00       50.0   

      CustomerID Country  
1423     14911.0    EIRE

In [71]:
df[df['StockCode'] == 'AMAZONFEE'].head(2)

InvoiceNo  StockCode Description  Quantity         InvoiceDate  \
14514   C537600  AMAZONFEE  AMAZON FEE        -1 2010-12-07 12:41:00   
15016   C537630  AMAZONFEE  AMAZON FEE        -1 2010-12-07 15:04:00   

       UnitPrice  CustomerID         Country  
14514       1.00         NaN  United Kingdom  
15016   13541.33         NaN  United Kingdom

In [72]:
df[df['StockCode'] == 'DCGSSGIRL'].head(1)

InvoiceNo  StockCode      Description  Quantity         InvoiceDate  \
84017    543358  DCGSSGIRL  GIRLS PARTY BAG         3 2011-02-07 14:04:00   

       UnitPrice  CustomerID         Country  
84017       3.29         NaN  United Kingdom

In [61]:
df[df['StockCode'] == 'BANK CHARGES'].head(1)

InvoiceNo     StockCode   Description  Quantity         InvoiceDate  \
4406    536779  BANK CHARGES  Bank Charges         1 2010-12-02 15:08:00   

      UnitPrice  CustomerID         Country  
4406       15.0     15823.0  United Kingdom

In [32]:
counter= 0
for descr in df['Description'].unique():
    if str(descr).islower():
        print(descr)
        counter += 1
print(counter)

nan
amazon
check
damages
faulty
amazon sales
reverse 21/5/10 adjustment
mouldy, thrown away.
found
counted
label mix up
samples/damages
thrown away
incorrectly made-thrown away.
showroom
wrongly sold as sets
dotcom sold sets
wrongly sold sets
? sold as sets?
?sold as sets?
damages/display
damaged stock
broken
throw away
wrong barcode (22467)
wrongly sold (22719) barcode
wrong barcode
barcode problem
?lost
thrown away-can't sell.
thrown away-can't sell
rcvd be air temp fix for dotcom sit
damages?
re dotcom quick fix.
sold in set?
cracked
sold as 22467
damaged
did  a credit  and did not tick ret
adjustment
returned
wrong code?
wrong code
adjust
crushed
damages/showroom etc
samples
mailout 
mailout
sold as set/6 by dotcom
wet/rusty
damages/dotcom?
on cargo order
smashed
reverse previous adjustment
wet damaged
missing
sold as set on dotcom
sold as set on dotcom and amazon
water damage
sold as set by dotcom
printing smudges/thrown away
to push order througha s stock was 
found some more on 

In [14]:
# data types
df.dtypes

InvoiceNo              object
StockCode              object
Description            object
Quantity                int64
InvoiceDate    datetime64[ns]
UnitPrice             float64
CustomerID            float64
Country                object
dtype: object

### Explore data
-----------

#### Description
* 1454 (0.3%) rows are empty

#### InvoiceDate

* extremply high number of purches happend between 2011-5 and 2011-07 wihch looks like error and should be checked

#### UnitPrice and Quantity

* mean values of `UnitPrice` and `Quantity` are 4.61 and 9.55 
* `df.describe` shows that 75% of values `UnitPrice` are concentrated near mean value but this column still has high Variance 9362 (but MAD=1.23 and Interquartile range is 2.82) as there are some points with high values for example max = 38970. Futhermore, 95-th percentile says that 95% of values are less then 9.95. Same happend on column Quantity. I guess it should be checked and clearned.
* 2515 prices are equal 0
#### CustomerID

* The most frequent customers are 17841 with 7983 number of records (1.5%), 14911 - 5903(1.1%) and 14096 - 5128 (0.9%)
* There are 4372 distinct customers

#### Country

* The most frequent countries are United Kingdom records 495478 (91.4%), Germany records 9495 (1.8%) and France 8557 (1.6%)
* 38 distinct countries





In [15]:
# convert data type
df.InvoiceDate = pd.to_datetime(df.InvoiceDate, format='%Y/%m/%d')
df.UnitPrice = df.UnitPrice.astype('float64')
df.Quantity = df.Quantity.astype('int64')

In [16]:
# describe
df.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

In [17]:
# histogram
sns.distplot(df.UnitPrice)

<AxesSubplot:xlabel='UnitPrice', ylabel='Density'>

In [18]:
plt.xlim(-100, 100)
sns.distplot(df.UnitPrice)

<AxesSubplot:xlabel='UnitPrice', ylabel='Density'>

In [19]:
# histogram
sns.distplot(df.Quantity)

<AxesSubplot:xlabel='Quantity', ylabel='Density'>

In [20]:
plt.xlim(-200, 200)
sns.distplot(df.Quantity)

<AxesSubplot:xlabel='Quantity', ylabel='Density'>

In [21]:
# random data wrangling
data = df.groupby('InvoiceDate', as_index=False).agg({'UnitPrice': np.sum, 'Quantity': np.sum})
sns.lineplot(data=data, x="InvoiceDate", y="UnitPrice")

<AxesSubplot:xlabel='Quantity', ylabel='Density'>

In [22]:
# pandas profiling
profile = ProfileReport(df, minimal=True)
profile.to_file(path + "data_understanding_report.html")
profile

Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

### Verify data quality
-------


#### StockCode
* Some records has value `POST D  DOT  M  BANK CHARGES  S  AMAZONFEE  m  DCGSSBOY  DCGSSGIRL  PADS 
B CRUK`

#### Description

* 1454 Description are missing
* There are records which used to describe some internal event but not sale example: 'mouldy, thrown away', 'damages', 'Given away' with 0.00 price. Posible solution remove all rows with lowercase letters
* Some records example 21518 and 38261 hava now product desctiption marked with '?' or NaN, price 0.00. Posible solution remove records with Nan,'?' or zero price

#### Quantity

* Not all records with min and max values are errors (example record 4287 with quantity -9360 is  returned products)

#### UnitPrice
* Some records has 0.00 value. Posible solution remove all columns with 0 price.
* There are 2 rows with negative price and one positive max value prices which marked with StockCode 'B' can be removed as it is Adjust bad debt

#### CustomerID
* 135080/406829 (24%) of fileds of customers are empty



In [23]:
# check duplicate row
df[df.duplicated(keep=False)]

InvoiceNo StockCode                       Description  Quantity  \
485       536409     22111      SCOTTIE DOG HOT WATER BOTTLE         1   
489       536409     22866     HAND WARMER SCOTTY DOG DESIGN         1   
494       536409     21866       UNION JACK FLAG LUGGAGE TAG         1   
517       536409     21866       UNION JACK FLAG LUGGAGE TAG         1   
521       536409     22900   SET 2 TEA TOWELS I LOVE LONDON          1   
...          ...       ...                               ...       ...   
541675    581538     22068       BLACK PIRATE TREASURE CHEST         1   
541689    581538     23318    BOX OF 6 MINI VINTAGE CRACKERS         1   
541692    581538     22992            REVOLVER WOODEN RULER          1   
541699    581538     22694                      WICKER STAR          1   
541701    581538     23343      JUMBO BAG VINTAGE CHRISTMAS          1   

               InvoiceDate  UnitPrice  CustomerID         Country  
485    2010-12-01 11:45:00       4.95     17908.0  United Kingdom  
489    2010-12-01 11:45:00       2.10     17908.0  United Kingdom  
494    2010-12-01 11:45:00       1.25     17908.0  United Kingdom  
517    2010-12-01 11:45:00       1.25     17908.0  United Kingdom  
521    2010-12-01 11:45:00       2.95     17908.0  United Kingdom  
...                    ...        ...         ...             ...  
541675 2011-12-09 11:34:00       0.39     14446.0  United Kingdom  
541689 2011-12-09 11:34:00       2.49     14446.0  United Kingdom  
541692 2011-12-09 11:34:00       1.95     14446.0  United Kingdom  
541699 2011-12-09 11:34:00       2.10     14446.0  United Kingdom  
541701 2011-12-09 11:34:00       2.08     14446.0  United Kingdom  

[10147 rows x 8 columns]

In [24]:
# check outlier
from scipy import stats
z_thr = 3.0
df[(np.abs(stats.zscore(df.select_dtypes(include=int))) > z_thr).any(axis=1)]

InvoiceNo StockCode                          Description  Quantity  \
4287     C536757     84347  ROTATING SILVER ANGELS T-LIGHT HLDR     -9360   
4850      536809     84950       ASSORTED COLOUR T-LIGHT HOLDER      1824   
4945      536830     84077    WORLD WAR 2 GLIDERS ASSTD DESIGNS      2880   
4946      536830     21915               RED  HARMONICA IN BOX       1400   
6365      536890    17084R                ASSORTED INCENSE PACK      1440   
9745      537214     17003                  BROCADE RING PURSE        720   
16427     537657     22189              CREAM HEART CARD HOLDER       972   
16428     537657     22188              BLACK HEART CARD HOLDER       972   
16435     537659     22189              CREAM HEART CARD HOLDER      1008   
16436     537659     22188              BLACK HEART CARD HOLDER      1008   
19149     537841     16014          SMALL CHINESE STYLE SCISSOR      1000   
19871     537899     22328   ROUND SNACK BOXES SET OF 4 FRUITS       1488   
19980     537981     22492              MINI PAINT SET VINTAGE       1394   
21518     538090     20956                                    ?      -723   
25920     538420     17096    ASSORTED LAQUERED INCENSE HOLDERS      1728   
32671     539101     22693   GROW A FLYTRAP OR SUNFLOWER IN TIN      2400   
38261     539494     21479                                    ?       752   
38463     539578     84992       72 SWEETHEART FAIRY CAKE CASES       720   
38468     539578     21212      PACK OF 72 RETROSPOT CAKE CASES       960   
40060     539721     22616           PACK OF 12 LONDON TISSUES       1296   
44863     540241     35957                                  NaN      -939   
44906     540245     17003                  BROCADE RING PURSE        720   
50849     540564     22617                 mouldy, thrown away.     -2600   
51049     540638    72038P                              damages      -990   
52131     540689     21733     RED HANGING HEART T-LIGHT HOLDER       860   
52132     540689    85123A   WHITE HANGING HEART T-LIGHT HOLDER      1010   
52133     540689     22469                HEART OF WICKER SMALL      1356   
52134     540689     22470                HEART OF WICKER LARGE      1284   
52262     540699      POST                                  NaN      1000   
52709     540815    85123A   WHITE HANGING HEART T-LIGHT HOLDER      1930   
52710     540815     21175          GIN + TONIC DIET METAL SIGN      2000   
52711     540815     21108   FAIRY CAKE FLANNEL ASSORTED COLOUR      3114   
52771     540818    47556B                 TEA TIME TEA TOWELS       1300   
52772     540818     48185                   DOORMAT FAIRY CAKE       670   
61619     541431     23166       MEDIUM CERAMIC TOP STORAGE JAR     74215   
61624    C541433     23166       MEDIUM CERAMIC TOP STORAGE JAR    -74215   
61671     541479     17003                  BROCADE RING PURSE        720   
63444     541570     21212      PACK OF 72 RETROSPOT CAKE CASES      1440   
65063     541685     22351                           Given away     -1400   
71761     542207     17003                  BROCADE RING PURSE        720   
71767     542208     17003                  BROCADE RING PURSE        720   
74614     542504     37413                                  NaN      5568   
74615     542505    79063D                                  NaN      2560   
77047     542690     70006             LOVE HEART POCKET WARMER       900   
80665     543051    79062D                                  NaN       960   
80742     543057     84077    WORLD WAR 2 GLIDERS ASSTD DESIGNS      2592   
80983     543099     22985         WRAP, BILLBOARD FONTS DESIGN       800   
82205     543192     15036            ASSORTED COLOURS SILK FAN      1200   
82208     543192     22616           PACK OF 12 LONDON TISSUES        864   
82794     543257    84611B                          thrown away     -1430   
82795     543258    84611B                                  NaN      1287   
865

In [25]:
#Check stockes code values
import re
def has_symbols(x):
    if len(re.findall(r'[A-Za-z]',str(x)))>0 and not re.findall(r'[0-9]',str(x)):
        return True
    else:
        return False
for elem in df[df['StockCode'].apply(has_symbols)]['StockCode'].unique():
    print(elem,' ')

for elem in df[df['StockCode'].apply(has_symbols)]['StockCode'].unique():
    display((df[df['StockCode']==elem]).head(1))
#df[df['StockCode'].apply(has_symbols)]

POST  
D  
DOT  
M  
BANK CHARGES  
S  
AMAZONFEE  
m  
DCGSSBOY  
DCGSSGIRL  
PADS  
B  
CRUK  


InvoiceNo StockCode Description  Quantity         InvoiceDate  UnitPrice  \
45    536370      POST     POSTAGE         3 2010-12-01 08:45:00       18.0   

    CustomerID Country  
45     12583.0  France

InvoiceNo StockCode Description  Quantity         InvoiceDate  UnitPrice  \
141   C536379         D    Discount        -1 2010-12-01 09:41:00       27.5   

     CustomerID         Country  
141     14527.0  United Kingdom

InvoiceNo StockCode     Description  Quantity         InvoiceDate  \
1814    536544       DOT  DOTCOM POSTAGE         1 2010-12-01 14:32:00   

      UnitPrice  CustomerID         Country  
1814     569.77         NaN  United Kingdom

InvoiceNo StockCode Description  Quantity         InvoiceDate  UnitPrice  \
2239    536569         M      Manual         1 2010-12-01 15:35:00       1.25   

      CustomerID         Country  
2239     16274.0  United Kingdom

InvoiceNo     StockCode   Description  Quantity         InvoiceDate  \
4406    536779  BANK CHARGES  Bank Charges         1 2010-12-02 15:08:00   

      UnitPrice  CustomerID         Country  
4406       15.0     15823.0  United Kingdom

InvoiceNo StockCode Description  Quantity         InvoiceDate  \
14436   C537581         S     SAMPLES        -1 2010-12-07 12:03:00   

       UnitPrice  CustomerID         Country  
14436      12.95         NaN  United Kingdom

InvoiceNo  StockCode Description  Quantity         InvoiceDate  \
14514   C537600  AMAZONFEE  AMAZON FEE        -1 2010-12-07 12:41:00   

       UnitPrice  CustomerID         Country  
14514        1.0         NaN  United Kingdom

InvoiceNo StockCode Description  Quantity         InvoiceDate  \
40383    539736         m      Manual         1 2010-12-21 15:18:00   

       UnitPrice  CustomerID         Country  
40383       2.55         NaN  United Kingdom

InvoiceNo StockCode     Description  Quantity         InvoiceDate  \
84016    543358  DCGSSBOY  BOYS PARTY BAG         1 2011-02-07 14:04:00   

       UnitPrice  CustomerID         Country  
84016       3.29         NaN  United Kingdom

InvoiceNo  StockCode      Description  Quantity         InvoiceDate  \
84017    543358  DCGSSGIRL  GIRLS PARTY BAG         3 2011-02-07 14:04:00   

       UnitPrice  CustomerID         Country  
84017       3.29         NaN  United Kingdom

InvoiceNo StockCode                 Description  Quantity  \
157195    550193      PADS  PADS TO MATCH ALL CUSHIONS         1   

               InvoiceDate  UnitPrice  CustomerID         Country  
157195 2011-04-15 09:27:00      0.001     13952.0  United Kingdom

InvoiceNo StockCode      Description  Quantity         InvoiceDate  \
299982   A563185         B  Adjust bad debt         1 2011-08-12 14:50:00   

        UnitPrice  CustomerID         Country  
299982   11062.06         NaN  United Kingdom

InvoiceNo StockCode      Description  Quantity         InvoiceDate  \
317508   C564763      CRUK  CRUK Commission        -1 2011-08-30 10:49:00   

        UnitPrice  CustomerID         Country  
317508        1.6     14096.0  United Kingdom

In [47]:
import re
def has_lower_case(x):
    if  x.isna():
        return False
    if len(re.findall(r'[a-z]',str(x)))>len(re.findall(r'[A-Z]',str(x))):
        print(x)
        return True
    else:
        return False
df[df['Description'].apply(has_lower_case)]

SyntaxError: invalid syntax (3652667390.py, line 3)

In [26]:
#Check negative price

#Check meaning of 'B' StockCode
print('Negative Unit price')
display(df[df['UnitPrice']<0])
print('StockCode B')
df[df['StockCode']=='B']

Negative Unit price


InvoiceNo StockCode      Description  Quantity         InvoiceDate  \
299983   A563186         B  Adjust bad debt         1 2011-08-12 14:51:00   
299984   A563187         B  Adjust bad debt         1 2011-08-12 14:52:00   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

StockCode B


InvoiceNo StockCode      Description  Quantity         InvoiceDate  \
299982   A563185         B  Adjust bad debt         1 2011-08-12 14:50:00   
299983   A563186         B  Adjust bad debt         1 2011-08-12 14:51:00   
299984   A563187         B  Adjust bad debt         1 2011-08-12 14:52:00   

        UnitPrice  CustomerID         Country  
299982   11062.06         NaN  United Kingdom  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

In [27]:
# check missing data
df.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

In [28]:
# check missing data for Description
df[df['Description'].isnull()]

InvoiceNo StockCode Description  Quantity         InvoiceDate  \
622       536414     22139         NaN        56 2010-12-01 11:52:00   
1970      536545     21134         NaN         1 2010-12-01 14:32:00   
1971      536546     22145         NaN         1 2010-12-01 14:33:00   
1972      536547     37509         NaN         1 2010-12-01 14:33:00   
1987      536549    85226A         NaN         1 2010-12-01 14:34:00   
...          ...       ...         ...       ...                 ...   
535322    581199     84581         NaN        -2 2011-12-07 18:26:00   
535326    581203     23406         NaN        15 2011-12-07 18:31:00   
535332    581209     21620         NaN         6 2011-12-07 18:35:00   
536981    581234     72817         NaN        27 2011-12-08 10:33:00   
538554    581408     85175         NaN        20 2011-12-08 14:06:00   

        UnitPrice  CustomerID         Country  
622           0.0         NaN  United Kingdom  
1970          0.0         NaN  United Kingdom  
1971          0.0         NaN  United Kingdom  
1972          0.0         NaN  United Kingdom  
1987          0.0         NaN  United Kingdom  
...           ...         ...             ...  
535322        0.0         NaN  United Kingdom  
535326        0.0         NaN  United Kingdom  
535332        0.0         NaN  United Kingdom  
536981        0.0         NaN  United Kingdom  
538554        0.0         NaN  United Kingdom  

[1454 rows x 8 columns]

In [34]:
# check missing data for Description
df[df['CustomerID'].isnull()]

InvoiceNo StockCode                      Description  Quantity  \
622       536414     22139                              NaN        56   
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786               POLKADOT RAIN HAT          4   
1446      536544     21787            RAIN PONCHO RETROSPOT         2   
...          ...       ...                              ...       ...   
541536    581498    85099B          JUMBO BAG RED RETROSPOT         5   
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE         4   
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN         1   
541539    581498     85174                S/4 CACTI CANDLES         1   
541540    581498       DOT                   DOTCOM POSTAGE         1   

               InvoiceDate  UnitPrice  CustomerID         Country  
622    2010-12-01 11:52:00       0.00         NaN  United Kingdom  
1443   2010-12-01 14:32:00       2.51         NaN  United Kingdom  
1444   2010-12-01 14:32:00       2.51         NaN  United Kingdom  
1445   2010-12-01 14:32:00       0.85         NaN  United Kingdom  
1446   2010-12-01 14:32:00       1.66         NaN  United Kingdom  
...                    ...        ...         ...             ...  
541536 2011-12-09 10:26:00       4.13         NaN  United Kingdom  
541537 2011-12-09 10:26:00       4.13         NaN  United Kingdom  
541538 2011-12-09 10:26:00       4.96         NaN  United Kingdom  
541539 2011-12-09 10:26:00      10.79         NaN  United Kingdom  
541540 2011-12-09 10:26:00    1714.17         NaN  United Kingdom  

[135080 rows x 8 columns]

In [29]:
# check number of unique value
df.nunique()

InvoiceNo      25900
StockCode       4070
Description     4223
Quantity         722
InvoiceDate    23260
UnitPrice       1630
CustomerID      4372
Country           38
dtype: int64

In [30]:
df.columns[0:3]

Index(['InvoiceNo', 'StockCode', 'Description'], dtype='object')

In [31]:
type(df['InvoiceNo'][0])

int

In [32]:
# check unique value
#columns with str type was removed
for feature in df.columns.delete([0,1,2]):
  print('------' + feature + '------')
  print(np.sort(df[feature].unique()))
  print()


------Quantity------
[-80995 -74215  -9600  -9360  -9058  -5368  -4830  -3667  -3167  -3114
  -3100  -3000  -2880  -2834  -2618  -2600  -2472  -2376  -2003  -2000
  -1930  -1897  -1681  -1671  -1560  -1515  -1512  -1510  -1479  -1440
  -1430  -1400  -1350  -1300  -1296  -1284  -1277  -1206  -1200  -1158
  -1131  -1128  -1121  -1100  -1092  -1060  -1050  -1000   -990   -975
   -967   -960   -939   -905   -900   -864   -840   -828   -800   -786
   -756   -752   -750   -741   -723   -720   -701   -696   -690   -682
   -675   -674   -670   -664   -657   -635   -624   -620   -618   -600
   -590   -576   -553   -550   -541   -530   -524   -504   -500   -484
   -480   -472   -468   -458   -455   -450   -443   -434   -432   -430
   -428   -420   -413   -408   -400   -398   -391   -390   -384   -380
   -365   -360   -355   -346   -345   -343   -342   -338   -337   -335
   -334   -327   -324   -323   -320   -318   -313   -312   -310   -306
   -304   -300   -290   -288   -280   -278   -276   -275

In [33]:
#show columns 0,1,2
for feature in df.columns[0:3]:
  print('------' + feature + '------')
  print(np.sort(df[feature].astype('str').unique()))
  print()

------InvoiceNo------
['536365' '536366' '536367' ... 'C581499' 'C581568' 'C581569']

------StockCode------
['10002' '10080' '10120' ... 'gift_0001_40' 'gift_0001_50' 'm']

------Description------
[' 4 PURPLE FLOCK DINNER CANDLES' " 50'S CHRISTMAS GIFT BAG LARGE"
 ' DOLLY GIRL BEAKER' ... 'wrongly sold (22719) barcode'
 'wrongly sold as sets' 'wrongly sold sets']

